src/integrating.jl

```diff
@@ -0,0 +1,105 @@
+"""
+    gauss_weights::Vector{Vector{Float64}}
+
+Precomputed Gaussian node weights up to degree 2*10-1 = 19.
+Computed using FastGaussQuadrature.jl with the command `[gausslegendre(i)[2] for i in 1:10]`
+"""
+gauss_weights = [[2.0], 
+                 [1.0, 1.0], 
+                 [0.5555555555555556, 0.8888888888888888, 0.5555555555555556], 
+                 [0.34785484513745385, 0.6521451548625462, 0.6521451548625462, 0.34785484513745385], 
+                 [0.23692688505618908, 0.47862867049936647, 0.5688888888888889, 0.47862867049936647, 0.23692688505618908], 
+                 [0.17132449237917025, 0.3607615730481385, 0.46791393457269126, 0.46791393457269126, 0.3607615730481385, 0.17132449237917025], 
+                 [0.1294849661688702, 0.2797053914892766, 0.3818300505051189, 0.4179591836734694, 0.3818300505051189, 0.2797053914892766, 0.1294849661688702], 
+                 [0.10122853629037676, 0.22238103445337445, 0.31370664587788744, 0.36268378337836193, 0.36268378337836193, 0.31370664587788744, 0.22238103445337445, 0.10122853629037676], 
+                 [0.08127438836157437, 0.18064816069485742, 0.2606106964029354, 0.31234707704000275, 0.3302393550012598, 0.31234707704000275, 0.2606106964029354, 0.18064816069485742, 0.08127438836157437], 
+                 [0.06667134430868821, 0.14945134915058056, 0.21908636251598207, 0.2692667193099965, 0.2955242247147529, 0.2955242247147529, 0.2692667193099965, 0.21908636251598207, 0.14945134915058056, 0.06667134430868821]]
+
+
+"""
+    IntegrandValues{integrandType}
+
+A struct used to save values of the integrand values in `integrand::Vector{integrandType}`.
+"""
+struct IntegrandValues{integrandType}
+    integrand::Vector{integrandType}
+end
+
+"""
+    IntegrandValues(integrandType::DataType)
+
+Return `IntegrandValues{integrandType}` with empty storage vectors.
+"""
+function IntegrandValues(::Type{integrandType}) where {integrandType}
+    IntegrandValues{integrandType}(Vector{integrandType}())
+end
+
+function Base.show(io::IO, integrand_values::IntegrandValues)
+    integrandType = eltype(integrand_values.integrand)
+    print(io, "IntegrandValues{integrandType=", integrandType, "}",
+          "\nintegrand:\n", integrand_values.integrand)
+end
+
+mutable struct SavingIntegrandAffect{IntegrandFunc, integrandType}
+    integrand_func::IntegrandFunc
+    integrand_values::IntegrandValues{integrandType}
+end
+
+function (affect!::SavingIntegrandAffect)(integrator)
+    n = div(OrdinaryDiffEq.alg_order(integrator.alg) + 1,2)
+    integral = zeros(eltype(eltype(affect!.integrand_values.integrand)),length(integrator.p))
+    for i in 1:n
+        t_temp = ((integrator.t-integrator.tprev)/2)*gauss_points[n][i]+(integrator.t+integrator.tprev)/2
+        integral .+= gauss_weights[n][i]*affect!.integrand_func(integrator(t_temp), t_temp, integrator)
```

Comment:
One performance optimization: if `isinplace(integrator.sol.prob)`, then we may want to have a cache vector and do `integrator(cache,t_temp)` in-place into the cache, and then `@. integral += gauss_weights[n][i]*affect!.integrand_func(cache, t_temp, integrator)`. But also, we may want to allow for the `integrand_func` itself to be `integrand_func!(obs,u,t,integrator)` that writes into `obs` and make this:

```julia
is inplace(integrator.sol.prob)
  integrator(cache,t_temp)`
  affect!.integrand_func(obs, t_temp, integrator)
  @. integral += gauss_weights[n][i]*obs
else
  integral .+= gauss_weights[n][i]*affect!.integrand_func(integrator(t_temp), t_temp, integrator)
end
```

That way it can be fully non-allocating.